### PR TITLE
Allow enabling stricter task validation

### PR DESCRIPTION
**Pull request by wolfs**
So all tasks can be validated as if they were cacheable tasks.

### CODE REVIEW

File: subprojects/plugin-development/src/main/java/org/gradle/plugin/devel/tasks/ValidateTaskProperties.java [hunk 1]
```diff
@@ -105,14 +110,24 @@
...
-    private FileCollection classpath;
-    private RegularFileProperty outputFile = getProject().getObjects().fileProperty();
+@SuppressWarnings("deprecation")
+public class ValidateTaskProperties extends ConventionTask implements VerificationTask, org.gradle.plugin.devel.tasks.internal.ValidateTaskPropertiesBackwardsCompatibleAdapter {
+    private final ConfigurableFileCollection classes;
+    private final ConfigurableFileCollection classpath;
+    private final RegularFileProperty outputFile;
+    private final Property<Boolean> enableStricterValidation;
     private boolean ignoreFailures;
     private boolean failOnWarning;
```
Comment (lptr) [inline subprojects/plugin-development/src/main/java/org/gradle/plugin/devel/tasks/ValidateTaskProperties.java]: We should change these to properties as well.

File: subprojects/plugin-development/src/main/java/org/gradle/plugin/devel/tasks/ValidateTaskProperties.java [hunk 2]
```diff
@@ -105,14 +110,24 @@
  * @since 3.0
  */
 @CacheableTask
-@SuppressWarnings("WeakerAccess")
-public class ValidateTaskProperties extends ConventionTask implements VerificationTask {
-    private FileCollection classes;
-    private FileCollection classpath;
-    private RegularFileProperty outputFile = getProject().getObjects().fileProperty();
+@SuppressWarnings("deprecation")
```
Comment (lptr) [inline subprojects/plugin-development/src/main/java/org/gradle/plugin/devel/tasks/ValidateTaskProperties.java]: Why is this needed?

File: subprojects/plugin-development/src/main/java/org/gradle/plugin/devel/tasks/ValidateTaskProperties.java [hunk 3]
```diff
@@ -268,35 +283,42 @@ public void setIgnoreFailures(boolean ignoreFailures) {
...
 
     /**
      * Sets the classpath used to load the classes under validation.
+     *
+     * @deprecated Use {@link #getClasspath()}.setFrom instead.
      */
+    @Deprecated
     public void setClasspath(FileCollection classpath) {
-        this.classpath = classpath;
+        this.classpath.setFrom(classpath);
```
Comment (lptr) [inline subprojects/plugin-development/src/main/java/org/gradle/plugin/devel/tasks/ValidateTaskProperties.java]: This should nag the user.

File: subprojects/plugin-development/src/main/java/org/gradle/plugin/devel/tasks/ValidateTaskProperties.java [hunk 4]
```diff
@@ -268,35 +283,42 @@ public void setIgnoreFailures(boolean ignoreFailures) {
...
     /**
      * Sets the classes to validate.
      *
      * @since 4.0
+     * @deprecated Use {@link #getClasses()}.setFrom instead.
      */
+    @Deprecated
     public void setClasses(FileCollection classes) {
-        this.classes = classes;
+        this.classes.setFrom(classes);
```
Comment (lptr) [inline subprojects/plugin-development/src/main/java/org/gradle/plugin/devel/tasks/ValidateTaskProperties.java]: This should nag the user.

File: subprojects/plugin-development/src/main/java/org/gradle/plugin/devel/tasks/ValidateTaskProperties.java [hunk 5]
```diff
@@ -105,14 +110,24 @@
...
     private boolean ignoreFailures;
     private boolean failOnWarning;
 
+    @Inject
+    public ValidateTaskProperties(ObjectFactory objects, ProjectLayout layout) {
+        this.classes = layout.configurableFiles();
+        this.classpath = layout.configurableFiles();
+        this.outputFile = objects.fileProperty();
+        this.enableStricterValidation = objects.property(Boolean.class);
+        this.enableStricterValidation.set(Boolean.FALSE);
```
Comment (lptr) [inline subprojects/plugin-development/src/main/java/org/gradle/plugin/devel/tasks/ValidateTaskProperties.java]: Why not just use `false` here?

File: subprojects/plugin-development/src/main/java/org/gradle/plugin/devel/tasks/ValidateTaskProperties.java (hunk 2 above)
Comment (wolfs) [reply subprojects/plugin-development/src/main/java/org/gradle/plugin/devel/tasks/ValidateTaskProperties.java]: Since I marked the interface which causes to create the bridge methods as deprecated.

File: subprojects/plugin-development/src/main/java/org/gradle/plugin/devel/tasks/ValidateTaskProperties.java (hunk 4 above)
Comment (wolfs) [reply subprojects/plugin-development/src/main/java/org/gradle/plugin/devel/tasks/ValidateTaskProperties.java]: Nagging.

File: subprojects/plugin-development/src/main/java/org/gradle/plugin/devel/tasks/ValidateTaskProperties.java (hunk 3 above)
Comment (wolfs) [reply subprojects/plugin-development/src/main/java/org/gradle/plugin/devel/tasks/ValidateTaskProperties.java]: Nagging.

File: subprojects/plugin-development/src/main/java/org/gradle/plugin/devel/tasks/ValidateTaskProperties.java (hunk 1 above)
Comment (wolfs) [reply subprojects/plugin-development/src/main/java/org/gradle/plugin/devel/tasks/ValidateTaskProperties.java]: I ended up using `boolean` for all the properties. If we want to use `Property<Boolean>`, then:
- This is a breaking change
- We cannot implement `VerificationTask` any more.

Or I would need to come up with some new property names for the `Property<Boolean>` setters...

**Comment by wolfs** [comment]
@lptr PTAL!

File: subprojects/plugin-development/src/main/java/org/gradle/plugin/devel/tasks/ValidateTaskProperties.java (hunk 2 above)
Comment (lptr) [reply subprojects/plugin-development/src/main/java/org/gradle/plugin/devel/tasks/ValidateTaskProperties.java]: Does this mean that if I use `ValidateTaskProperties` in my code it will be marked as deprecated, too?

Comment (wolfs) [reply subprojects/plugin-development/src/main/java/org/gradle/plugin/devel/tasks/ValidateTaskProperties.java]: I tried and I didn't get any deprecation warnings. Not for extending the class nor for calling `getClasses()`.

**Comment by wolfs** [comment]
@lptr I added release notes. PTAL!

File: subprojects/docs/src/docs/release/notes.md [hunk 6]
```diff
@@ -10,6 +10,12 @@ Add-->
 ### Example new and noteworthy
 -->
 
+### Stricter validation with validateTaskProperties
```
Comment (lptr) [inline subprojects/docs/src/docs/release/notes.md]: We should either mention the task name (`validateTaskProperties`) or the task type (`ValidateTaskProperties`) consistently I think. The different casing is confusing.

Comment (lptr) [reply subprojects/docs/src/docs/release/notes.md]: Should also escape the name with backticks (``).

File: subprojects/docs/src/docs/release/notes.md [hunk 7]
```diff
@@ -40,7 +40,7 @@ The following are the newly deprecated items in this Gradle release. If you have
 ### Example deprecation
 -->
 
-### Setters for `classes` and `classpath` on `ValidateTaskProperties`
+### Setters for `classes` and `classpath` on `validateTaskProperties`
```
Comment (lptr) [inline subprojects/docs/src/docs/release/notes.md]: Not: I think here we should use the type name.
